File: csrc/390.cpp
//
// Created by linkinpark213 on 5/29/19.
//

#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int lastRemaining(int n) {
        if (n == 1) return 1;
        else if (n <= 3) return 2;
        else {
            if ((n / 2) % 2 == 1) {
                return 4 * lastRemaining((n - 2) / 4);
            } else {
                return 4 * lastRemaining(n / 4) - 2;
        }
        }
    }
};

int main() {
    Solution solution;
    cout << solution.lastRemaining(9);
}

File: pysrc/390.py
class Solution:
    def lastRemainingLeft(self, n: int) -> int:
        if n == 1:
            return 1
        elif n == 2:
            return 2
        return 2 * self.lastRemainingRight(n // 2)

    def lastRemainingRight(self, n: int) -> int:
        if n == 1:
            return 1
        elif n == 2:
            return 1
        if n % 2 == 0:
            return 2 * self.lastRemainingLeft(n // 2) - 1
        elif n % 2 == 1:
            return 2 * self.lastRemainingLeft(n // 2)

    def lastRemaining(self, n: int) -> int:
        if n < 2:
            return n
        return self.lastRemainingLeft(n)


if __name__ == '__main__':
    solution = Solution()
    print(solution.lastRemainingLeft(9))
